### PR TITLE
fixes to colonful `reshape`

### LINKED ISSUE #54245: colonful `reshape` may spuriously throw `DivideError`
```julia-repl
julia> reshape([], (1, :)) == reshape([], (1, 0))
true

julia> reshape([], (1, :)) == reshape([], (1, 0)) == reshape([], (0, :))  # shouldn't throw
ERROR: DivideError: integer division error
Stacktrace:
 [1] div
   @ ./int.jl:295 [inlined]
 [2] divrem
   @ ./div.jl:203 [inlined]
 [3] divrem
   @ ./div.jl:181 [inlined]
 [4] _reshape_uncolon
   @ ./reshapedarray.jl:138 [inlined]
 [5] reshape(parent::Vector{Any}, dims::Tuple{Int64, Colon})
   @ Base ./reshapedarray.jl:129
 [6] top-level scope
   @ REPL[2]:1

julia> reshape([1], (0, :))  # should throw, but not `DivideError`
ERROR: DivideError: integer division error
Stacktrace:
 [1] div
   @ ./int.jl:295 [inlined]
 [2] divrem
   @ ./div.jl:203 [inlined]
 [3] divrem
   @ ./div.jl:181 [inlined]
 [4] _reshape_uncolon
   @ ./reshapedarray.jl:138 [inlined]
 [5] reshape(parent::Vector{Int64}, dims::Tuple{Int64, Colon})
   @ Base ./reshapedarray.jl:129
 [6] top-level scope
   @ REPL[2]:1

julia> b = prevpow(2, typemax(Int))
4611686018427387904

julia> iszero(b*b)  # `b` is a square root of zero, thus a zero divisor
true

julia> reshape([1], (b, :, b))  # should throw, but not `DivideError`
ERROR: DivideError: integer division error
Stacktrace:
 [1] div
   @ ./int.jl:295 [inlined]
 [2] divrem
   @ ./div.jl:203 [inlined]
 [3] divrem
   @ ./div.jl:181 [inlined]
 [4] _reshape_uncolon
   @ ./reshapedarray.jl:138 [inlined]
 [5] reshape(parent::Vector{Int64}, dims::Tuple{Int64, Colon, Int64})
   @ Base ./reshapedarray.jl:129
 [6] top-level scope
   @ REPL[6]:1

julia> reshape([], (b, :, b))
ERROR: DivideError: integer division error
Stacktrace:
 [1] div
   @ ./int.jl:295 [inlined]
 [2] divrem
   @ ./div.jl:203 [inlined]
 [3] divrem
   @ ./div.jl:181 [inlined]
 [4] _reshape_uncolon
   @ ./reshapedarray.jl:138 [inlined]
 [5] reshape(parent::Vector{Any}, dims::Tuple{Int64, Colon, Int64})
   @ Base ./reshapedarray.jl:129
 [6] top-level scope
   @ REPL[4]:1
```

For the last example, the expected result depends on the resolution of #54244, but in any case it shouldn't throw `DivideError`.

I've got a fix, but waiting for a decision on #54244.

```julia-repl
julia> versioninfo()
Julia Version 1.12.0-DEV.387
Commit b5bfd83a3d0 (2024-04-22 13:12 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 8 × AMD Ryzen 3 5300U with Radeon Graphics
  WORD_SIZE: 64
  LLVM: libLLVM-16.0.6 (ORCJIT, znver2)
Threads: 1 default, 0 interactive, 1 GC (on 8 virtual cores)
Environment:
  JULIA_NUM_PRECOMPILE_TASKS = 3
  JULIA_PKG_PRECOMPILE_AUTO = 0
```

This is the affected code:

https://github.com/JuliaLang/julia/blob/8f6418ea507c2a478b185f6dee66d5877f75328d/base/reshapedarray.jl#L130-L141

### PR DESCRIPTION
**Pull request by nsajko**
Fixes #54245

### CODE REVIEW

**Review by nhz2 — APPROVED** [review]
LGTM

File: base/reshapedarray.jl [hunk 1]
```diff
@@ -132,12 +132,23 @@ reshape(parent::AbstractArray, dims::Tuple{Vararg{Union{Int,Colon}}}) = reshape(
...
+    sz, remainder = if iszero(len)
+        (0, 0)
+    else
+        let pr = Core.checked_dims(pre..., post...)  # safe product
+            if iszero(pr)
+                throw2(A, dims)
+            end
+            divrem(len, pr)
+        end
+    end::NTuple{2,Int}
```
[the remark this answers is not in the document]
Comment (nsajko) [reply base/reshapedarray.jl]: Fixed.

> BTW, do we really need these `typeassert`? It seems unlikely to bring inference improvement.

I see a type assert as a strictly good thing, as long as it's not causing a bug. If it's not helping the compiler's inference, the type assert should be optimized out.

**Comment by nsajko** [comment]
ping

**Comment by nsajko** [comment]
ping

**Review by oscardssmith — APPROVED** [review]

**Comment by jishnub** [comment]
The type assertions added here appear to have broken reshape using `BigInt`s. Are these essential for type-inference?
E.g. the following works on Julia v1.10, but is broken on nightly:
```julia
julia> using FillArrays

julia> reshape(Fill(1, 2), big(2), :)
2×1 Fill{Int64}, with entries equal to 1
```

**Pull request merged**